<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [1672]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [1673]:
# текст запроса
query_3_1 = f'''select
                count(*)
                from vacancies                
'''

In [1674]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
print(f'''Количество вакансий в базе: {df_3_1['count'][0]}''')

Количество вакансий в базе: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [1675]:
# текст запроса
query_3_2 = f'''select
                count(*)
                from employers                
'''

In [1676]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
print(f'''Количество работодателей в базе: {df_3_2['count'][0]}''')

Количество работодателей в базе: 23501


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [1677]:
# текст запроса
query_3_3 = f'''select
                count(*)
                from areas                
'''

In [1678]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
print(f'''Количество регионов: {df_3_3['count'][0]}''')

Количество регионов: 1362


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [1679]:
# текст запроса
query_3_4 = f'''select
                count(*)
                from industries                
'''

In [1680]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
print(f'''Количество сфер деятельности: {df_3_4['count'][0]}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество сфер деятельности: 294


***

# выводы по предварительному анализу данных
Всего в базе находится около 50 тыс вакансий. При этом работодателей примерно в два раза меньше.
Вакансии распределены по 1362 регионам и охватывают почти 300 различных сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [1681]:
# текст запроса
query_4_1 = f'''select
                    count(*) cnt,
                    a.name area
                from vacancies v
                join areas a on v.area_id = a.id
                group by a.id
                order by 1 DESC
                limit 5
'''

In [1682]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1
#print(f'''Количество сфер деятельности: {df_4_1}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [1683]:
# текст запроса
query_4_2 = f'''select
                    count(*)
                from vacancies
                where salary_from is not null or salary_to is not null
'''

In [1684]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
print(f'''Количество вакансий, в которых заполнено хотя бы одно поле с зарплатой: {df_4_2['count'][0]}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество вакансий, в которых заполнено хотя бы одно поле с зарплатой: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [1685]:
# текст запроса
query_4_3 = f'''select
                    Round(AVG(salary_from)) salary_from,
                    Round(AVG(salary_to)) salary_to
                from vacancies
'''

In [1686]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3
print(f'''Среднее значение для нижней границы зарплатной вилки: {df_4_3['salary_from'][0]} \nСреднее значение для верхней границы зарплатной вилки: {round(df_4_3['salary_to'][0])}''')
# round при выводе добавлен, чтобы убрать лишний ноль

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Среднее значение для нижней границы зарплатной вилки: 71065.0 
Среднее значение для верхней границы зарплатной вилки: 110537


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [1687]:
# текст запроса
query_4_4 = f'''select
                    schedule,
                    employment,
                    count(*) 
                from vacancies
                group by schedule, employment
                order by count desc
                
'''

In [1688]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [1689]:
# текст запроса
query_4_5 = f'''select
                    experience,
                    count(*) 
                from vacancies
                group by experience
                order by count
                
'''

In [1690]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
1) Ожидаемо, самое большое количество вакансий находится в Москве. При этом в следующих 4 городах, являющихся также столицами или крупными региональными центрами, количество вакансий примерно в 2 раза меньше, чем в Мосвке, и на треть меньше, но незначительно различается между собой (кроме Санкт-Петербурга, занимающего 2 место), что может означать примерно одинаковый размер рынка труда и развития экономики.
2) Всего около половины вакансий содержат информацию о предполагаемой зарплате. Это может означать высоко конкурентный рынок труда, где уровень зарплаты является предметом торга с соискателем. Либо высокую зависимость уровня зарплаты от конкретного профиля соискателя.
3) Средняя зарплатная вилка от 71 до 110 тыс рублей имеет более высокий уровень, чем медианная зарплата по РФ за 2020 год, которая составляла по данным Росстата 32 тыс руб. На основании этого можно сделать вывод, что на средние значения влияют аномально высокие зарплаты, либо работодатели предлагающие более низкие зарплаты не указывают значение в базе.
4) Классическое высокое сочетание полного дня и полной занятости - на первом месте с большим отрывом, на втором и четветом месте - удаленная работа, которая в последнее время начинает набирать популярность. Предложений по проектной работе и по стажировкам значительно меньше, чем остальных типов занятости, что может означать нацеленность работодателей на найм постоянных сотрудников.
5) Самый востребованный опыт - от года до 3 лет, на втором месте от 3 до 6 лет - это означает, что работодатели больше настроены на сотрудников, которые имеют некоторые знания, и которых можно "обучить под себя", при этом зарплатные ожидания таких соискателей не очень высокие, или сотрудников достаточно опытных, чтобы выполнять основные функции без дополнительных усилий со стороны работодателя. Соискатели без опыта требуются, обычно, на малоквалифицированные должности, либо, если работодатель готов инвестировать в обучение сотрудников с нуля. Реже всего требуются сотрудники с опытом более 6 лет, так как, скорее всего, речь идет о более высоких позициях (например, менеджерах), которых в компаниях меньше, чем средних и низких.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [1691]:
# текст запроса
query_5_1 = f'''select
                    e.name employer,
                    count(v.id) vacancies
                from vacancies v
                join employers e on v.employer_id = e.id
                group by employer
                order by vacancies desc
                limit 5
                
'''

In [1692]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer  vacancies
0         Яндекс       1933
1     Ростелеком        491
2       Тинькофф        444
3           СБЕР        428
4  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [1694]:
# текст запроса
query_5_2 = f'''select
                    a.name area,
                    count(e.id) employer,
                    count(v.id) vacancies    
                from areas a
                left join employers e on e.area = a.id
                left join vacancies v on a.id = v.area_id
                where v.name is null
                
                group by 1
                order by employer desc
                limit 1
                       
'''

In [1695]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  employer  vacancies
0  Россия       410          0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [1697]:
# текст запроса
query_5_3 = f'''select
                    e.name employer,
                    count(distinct v.area_id) areas_number    
                from employers e
                join vacancies v on e.id = v.employer_id            
                group by 1
                order by 2 desc
'''

In [1698]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer  areas_number
0                      Яндекс           181
1                  Ростелеком           152
2                  Спецремонт           116
3      Поляков Денис Иванович            88
4                   ООО ЕФИН             71
...                       ...           ...
14761                  UniSol             1
14762            UNISTORY LLC             1
14763                   UNIT6             1
14764     United Distribution             1
14765              UNITEL LLC             1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [1699]:
# текст запроса
query_5_4 = f'''select
                    count(e.id)  
                from employers e
                left join employers_industries ei on e.id = ei.employer_id
                where industry_id is null

'''

In [1700]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4
print(f'''Количество работодателей, у которых не указана сфера деятельности: {df_5_4['count'][0]}''')

Количество работодателей, у которых не указана сфера деятельности: 8419


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [1701]:
# текст запроса
query_5_5 = f'''select
                    e.name
                from employers e
                left join employers_industries ei on e.id = ei.employer_id
                group by e.name
                having count(industry_id) = 4
                order by e.name
                --limit 3
'''

In [1702]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5
print(f'''Компания, находящаяся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности: {df_5_5['name'][2]}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Компания, находящаяся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности: 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [1703]:
# текст запроса
query_5_6 = f'''select
                    count(e.name)
                from employers e
                left join employers_industries ei on e.id = ei.employer_id
                left join industries i on ei.industry_id = i.id 
                where i.name = 'Разработка программного обеспечения'
'''

In [1704]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
print(f'''Количество компаний, у которых указана сфера деятельности Разработка программного обеспечения: {df_5_6['count'][0]}''')

Количество компаний, у которых указана сфера деятельности Разработка программного обеспечения: 3553


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [1705]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
table1 = page.find('table', class_='standard')
headers = []
for i in table1.find_all('th'):
    title = i.text
    headers.append(title)

mydata = pd.DataFrame(columns = headers)

for j in table1.find_all('tr')[1:]:
    row_data = j.find_all('td')
    row = [i.text for i in row_data]
    length = len(mydata)
    mydata.loc[length] = row
    
value = str(tuple(mydata['Город']))

value = value.replace('[a]', '')
value

"('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')"

In [1706]:
query_5_7 = f'''select
                    --e.name employer,
                    a.name,
                    count(v.id) vacancies
                from vacancies v
                join employers e on v.employer_id = e.id
                join areas a on a.id = v.area_id
                group by e.name, a.name
                having e.name = 'Яндекс' and a.name in {value}
                
                union all
                
                select
                    'TOTAL',
                    count(v.id)
                from vacancies v
                join employers e on v.employer_id = e.id
                join areas a on a.id = v.area_id
                where e.name = 'Яндекс' and a.name in {value}
                order by vacancies
         
'''

In [1707]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  vacancies
0              Омск         21
1         Челябинск         22
2        Красноярск         23
3         Волгоград         24
4             Пермь         25
5            Казань         25
6    Ростов-на-Дону         25
7               Уфа         26
8            Самара         26
9         Краснодар         30
10          Воронеж         32
11      Новосибирск         35
12  Нижний Новгород         36
13     Екатеринбург         39
14  Санкт-Петербург         42
15           Москва         54
16            TOTAL        485

***

# выводы по анализу работодателей
1. На первых пяти местах по количеству вакансий находятся компании Яндекс, Ростелеком, Тиньков, СБЕР и Газпром нефть - которые являются большими и активно развивающими бизнес компаниями. При этом в Яндекс, занимающем первое место, размещено вакансий примерно в 4 раза больше, чем на втором месте. Что показывает значительный рост компании, что соотносится с трендом роста IT сферы.
2. Около 400 компаний, не разместивших вакансий, в качестве региона указывают Россия. Можно предположить, что это либо компании, работающие по всей территории, либо неполные данные
3. По количеству регионов, в которых размещены вакансии на первом и втором месте находятся Яндекс и Ростелеком, что коррелирует с общим количеством вакансий. При этом можно предположить, что вакансии Яндекс, скорее, относятся к удаленной работе, поэтому наблюдается такое количество регионов, в то время, как большое количество регионов для Ростелекома, скорее относится именно к сети офисов по России. В то же время стоит отметить, что на 3-5 местах находятся малоизвестные компании. Такое количество регионов для этих компаний можно связать либо с ошибкой при заполнении, либо с намеренным расширением списка регионов для привлечения кандидатов.
4. У около четверти работодателей (почти 8,5 тыс) из списка работодателей не указана сферы деятельности, можно предположить, что подходящей сферы не нашлось в списке, либо считают, что область деятельности не важна для поиска кандидата
5. При этом есть компании (около 1 тыс), которые указывают по 4 сферы деятельности.
6. Из оставшихся трех четвертей работодателей, указавших сферу деятельности, примерно у 10% указана сфера Разработка ПО. Что указывает на достаточно большую долю рынка и востребованость специалистов этого направления.
7. Из почти 2 тыс вакансий компании Яндекс, всего около черверти приходится на города-миллионники. Что подтверждает предположение об отсутствии концентрации вакансий в головном офисе в Москве (всего 54 вакансии, около 1% из городов миллионников), и стратегию привлечения сотрудников из разных регионов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [1708]:
# текст запроса
query_6_1 = f'''select
                    --name
                    count(*) 
                from vacancies
                where lower(name) like '%данн%' or lower(name) like '%data%'
        
'''

In [1709]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1
print(f'''Количество вакансий, имеющих отношение к данным: {df_6_1['count'][0]}''')

Количество вакансий, имеющих отношение к данным: 1771


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [1710]:
# текст запроса
query_6_2 = f'''select
                    --name,
                    --experience,
                    --employment
                    count(*) 
                from vacancies
                where (lower(name) like '%data scientist%' 
                    or lower(name) like '%data science%'
                    or lower(name) like '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or lower(name) like '%machine learning%'
                    or lower(name) like '%машинн%обучен%'
                    )
                    and
                    (lower(name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка')
                
        
'''

In [1711]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2
print(f'''Количество вакансий, подходящих для начинающего дата-сайентиста: {df_6_2['count'][0]}''')

Количество вакансий, подходящих для начинающего дата-сайентиста: 51


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [1712]:
# текст запроса
query_6_3 = f'''select
                    --name,
                    --key_skills
                    count(*) 
                from vacancies
                where (lower(name) like '%data scientist%' 
                    or lower(name) like '%data science%'
                    or lower(name) like '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or lower(name) like '%machine learning%'
                    or lower(name) like '%машинн%обучен%'
                    )
                    and
                    (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%')
                
        
'''

In [1713]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3
print(f'''Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres: {df_6_3['count'][0]}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres: 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [1714]:
# текст запроса
query_6_4 = f'''select
                    --name,
                    --key_skills
                    count(*) 
                from vacancies
                where (lower(name) like '%data scientist%' 
                    or lower(name) like '%data science%'
                    or lower(name) like '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or lower(name) like '%machine learning%'
                    or lower(name) like '%машинн%обучен%'
                    )
                    and
                    (lower(key_skills) like '%python%')
                
        
'''

In [1715]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4
print(f'''Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres: {df_6_4['count'][0]}''')

Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres: 351


c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [1716]:
# текст запроса
query_6_5 = f'''select
                    --name,
                    key_skills,
                    (length(key_skills) - length(replace(key_skills, '\t', '')) + 1) as skills_number
                    --count(*) 
                from vacancies
                where (lower(name) like '%data scientist%' 
                    or lower(name) like '%data science%'
                    or lower(name) like '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or lower(name) like '%machine learning%'
                    or lower(name) like '%машинн%обучен%'
                    )
'''

In [1717]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5
print(f'''Среднее количество ключевых навыков в вакансиях для DS: {round(df_6_5['skills_number'].mean(), 2)}''')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Среднее количество ключевых навыков в вакансиях для DS: 6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [1718]:
# текст запроса
query_6_6 = f'''select
                    experience,
                    round(avg((coalesce(salary_from, 0) + coalesce(salary_to,0)) / ((case when salary_from IS NULL then 0 else 1 end) + (case when salary_to IS NULL then 0 else 1 end)))) salary_avg
                from vacancies
                where (lower(name) like '%data scientist%' 
                    or lower(name) like '%data science%'
                    or lower(name) like '%исследователь данных%'
                    or (name like '%ML%' and name not like '%HTML%')
                    or lower(name) like '%machine learning%'
                    or lower(name) like '%машинн%обучен%'
                    )
                    and (salary_from is not null or salary_to is not null)
                group by 1 
'''

In [1719]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  salary_avg
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

# выводы по предметному анализу
1. Около 3% вакансий (1771) относятся к работе с данными. 
2. Количество вакансий, подходящих для начинающего дата-сайентиста - 51, что составлет примерно 0,1% от общего числа вакансий и около 10% от всех вакансий в дата сайенс (480). При этом общий уровень вакансий для новичков составляет порядка 16% от общего числа вакансий. Можно сделать вывод, что в целом рынок дата сайенс немного более насыщен новичками, чем остальные направления.
3. Из 480 вакансий, относящихся к DS в 201 требуются навыки SQL или postgres (примерно 42%). Можно сделать вывод, что знания баз данных является серьезным преимуществом при поиске работы в DS.
4. Вакансий в DS, требующих знания python 351, то есть около 73%. Можно сделать вывод, python является наиболее популярным языком программирования для DS
5. Среднее количество ключевых навыков в DS - 6,4, что примерно соответствует среднему значению по всем вакансиям. 
6. Средний уровень зарплаты для вакансий в DS начинается с примерно 75 тыс руб для новичков и далее растет почти в два раза за каждые 3 года опыта.

# Общий вывод по проекту

In [1720]:
# Проверка распределения вакансий по сферам деятельности
query_7_1 = f'''(select
                    i.name industry,
                    count(distinct v.id) vacancies_number
                from employers e
                join vacancies v on v.employer_id = e.id
                left join employers_industries ei on e.id = ei.employer_id
                left join industries i on ei.industry_id = i.id
                group by i.name
                order by count(v.id) desc
                limit 10)
                --where i.name = 'Разработка программного обеспечения'
                
                union ALL
                (select 'TOTAL', count(v.id)
                from vacancies v)
'''

In [1721]:
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


industry  vacancies_number
0                 Разработка программного обеспечения             12499
1   Системная интеграция,  автоматизации технологи...             11034
2                                                None             10084
3   Интернет-компания (поисковики, платежные систе...              6413
4                                                Банк              2742
5                                     Мобильная связь              1585
6                                  Кадровые агентства              1143
7                               Консалтинговые услуги              1137
8                                  Интернет-провайдер              1095
9                        Розничная сеть (продуктовая)              1034
10                                              TOTAL             49197

Вывод: количество вакансий, размещенных компаниями, указавших сферу деятельности "Разработка программного обеспечения", "Системная интеграция, автоматизации технологи..", "Интернет-компания (поисковики, платежные систе.." - в несколько раз выше, чем по другим сферам. Что показывает общую тенденцию развития данных сфер. Однако надо принимать во внимание, что в данной выборке количество вакансий, размещенной компанией, указавшей несколько сфер дейтельности, будет считаться для каждой указанной сферы.

In [1722]:
# Проводим сравнение средних зарплат в зависимости от опыта для DS и всех вакансий 
query_7_2 = f'''select
                    experience,
                    round(avg((coalesce(salary_from, 0) + coalesce(salary_to,0)) / ((case when salary_from IS NULL then 0 else 1 end) + (case when salary_to IS NULL then 0 else 1 end)))) salary_avg_all
                from vacancies
                where salary_from is not null or salary_to is not null
                group by 1
                order by 2
'''

In [1723]:
df_7_2 = pd.read_sql_query(query_7_2, connection)
df_7_2
pd.merge(df_7_2, df_6_6, how='outer', on='experience')

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  salary_avg_all  salary_avg
0           Нет опыта         40472.0     74643.0
1  От 1 года до 3 лет         76541.0    139675.0
2       От 3 до 6 лет        133211.0    243115.0
3         Более 6 лет        166256.0         NaN

Вывод: зарплаты для сферы DS примерно в 2 раза выше, чем средние по всем вакансиям

In [1724]:
# Проводим уточнение по типу графика вакансий Яндекс с учетом разбивки по городам-миллионникам (Cities) и остальным регионам (Regions)
query_7_3 = f'''(select
                    --e.name employer,
                    v.schedule,
                    case when a.name in {value} then 'Cities' else 'Regions' end area,
                    count(v.id) vacancies
                from vacancies v
                join employers e on v.employer_id = e.id
                join areas a on a.id = v.area_id
                where e.name = 'Яндекс'
                group by 1, 2
                order by 3 desc)
                
                union ALL
                (select '', 'TOTAL', count(v.id)
                from vacancies v
                join employers e on v.employer_id = e.id
                where e.name = 'Яндекс')
         
'''

In [1725]:
df_7_3 = pd.read_sql_query(query_7_3, connection)
df_7_3

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule     area  vacancies
0                      TOTAL       1933
1  Удаленная работа  Regions       1366
2  Удаленная работа   Cities        324
3       Полный день   Cities        117
4    Сменный график  Regions         38
5     Гибкий график  Regions         30
6     Гибкий график   Cities         25
7    Сменный график   Cities         19
8       Полный день  Regions         14

Вывод: Подтверждается предположение о том, что большая часть вакансий Яндекс относится к удаленной работе. БОлее того, из выборки видно, что более 70% вакансий относятся к удаленной работе именно в регионах.

In [1726]:
# Проверка наименований вакансий Спецремонт, для которого получено неожиданно большое количество регионов размещения вакансий (151)
query_7_4 = f'''(select
                    --e.name employer,
                    v.name,
                    v.schedule,
                    case when a.name in {value} then 'Cities' else 'Regions' end area,
                    count(v.id) vacancies_number
                from vacancies v
                join employers e on v.employer_id = e.id
                join areas a on a.id = v.area_id
                where e.name = 'Спецремонт'
                group by 1, 2, 3
                order by 1 desc)
                
                union ALL
                (select '', '', 'TOTAL', count(v.id)
                from vacancies v
                join employers e on v.employer_id = e.id
                where e.name = 'Спецремонт')
         
'''

In [1727]:
df_7_4 = pd.read_sql_query(query_7_3, connection)
df_7_4

c:\Users\Alexey\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name        schedule     area  vacancies_number
0                                                   TOTAL               151
1   Ученик компьютерного мастера   Гибкий график  Regions                 3
2   Ученик компьютерного мастера  Сменный график  Regions                25
3            Компьютерный Мастер   Гибкий график  Regions                16
4            Компьютерный Мастер     Полный день   Cities                 7
5            Компьютерный Мастер     Полный день  Regions                57
6            Компьютерный Мастер  Сменный график  Regions                 3
7            Компьютерный мастер   Гибкий график  Regions                17
8            Компьютерный мастер     Полный день   Cities                 1
9            Компьютерный мастер  Сменный график   Cities                 4
10           Компьютерный мастер  Сменный график  Regions                18

Вывод: С учето того, что получено всего два названия вакансий (2 варианта написания для Компьютерного мастера) для компании Спецремонт, можно предположить, что речь идет о небольшом количестве реальных вакансий, продублированных для разных регионов и графиков работы.

# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Всего в базе находится около 50 тыс вакансий. При этом работодателей примерно в два раза меньше.
Вакансии распределены по 1362 регионам и охватывают почти 300 различных сфер деятельности.
# выводы по детальному анализу вакансий
Общий рынок труда имеет тенденцию к концентрации в больших городах, так Москва занимает более 10% всех вакансий. При этом преобладает сочетание полного дня и полной занятости (около 70% всех вакансий), однако намечается тенденция на удаленную работу. Указанный средний уровень зарплат более чем в два раза превышает среднюю медианную зарплату РФ по данным Росстата, однако следует принимать во внимание, что примерно половина вакансий не содержат информацию о зарплате. Самым востребованным уровнем опыта является «от года до 3», что в целом отражает готовность организаций вкладывать время в развитие сотрудника, однако это компенсируется более низким уровнем зарплаты для таких вакансий. Реже всего требуются сотрудники с опытом более 6 лет, так как, скорее всего, речь идет о более высоких позициях (например, менеджерах), которых в компаниях меньше, чем средних и низких, также на такие позиции чаще берут сотрудников по рекомендациям.
# выводы по анализу работодателей
Анализ работодателей показывает, что на первых местах по количеству вакансий находятся компании в сфере IT, что также подтверждает дополнительный анализ по сферам деятельности. Одной из самых популярных сфер деятельности является «Разработка ПО» - около 10% всех компаний, указавших сферу деятельности. Яндекс является несомненным лидером по количеству размещенных вакансий – около 4% от всей базы. Более того, анализ регионального распределения вакансий Яндекс демонстрирует стратегию расширения географии за счет привлечения сотрудников в регионах на удаленную работу. 

# выводы по предметному анализу
 Около 3% вакансий относятся к работе с данными и около 1% - к DS. При этом к требуемым навыкам в DS в 42% относятся навыки SQL или postgres, а в 73% - знания python. Также стоит отметить, что количество вакансий для новичков в DS составляет 10% от общего числа вакансий в DS, тогда как среднее количество вакансий для новичков по всей базе составляет 16%. Можно сделать вывод, что в целом рынок DS немного более насыщен новичками, чем остальные направления. Средний уровень зарплаты для вакансий в DS начинается с примерно 75 тыс руб для новичков и далее растет почти в два раза за каждые 3 года опыта. При этом зарплаты для сферы DS примерно в 2 раза выше, чем средние по всем вакансиям

Для того, чтобы подтвердить тенденции развития рынка IT в целом и DS в частности, необходимо провести анализ изменения ситуации во времени.


In [ ]:
connection.close()